src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/Utils.java

```diff
@@ -85,4 +77,27 @@ public static VarHandle fixUpVarHandle(VarHandle handle) {
     private static MemoryAddressProxy filterAddress(MemoryAddress addr) {
         return (MemoryAddressImpl)addr;
     }
+
+    public static void checkRestrictedAccess() {
+        switch (foreignAccess) {
+            case "deny" -> throwIllegalAccessError(foreignAccess);
+            case "warn" -> System.err.println("WARNING: Accessing jdk.incubator.foreign.Foreign.");
+            case "debug" -> {
+                StringBuilder sb = new StringBuilder("DEBUG: Accessing jdk.incubator.foreign.Foreign.");
+                StackWalker.getInstance().walk(s -> {
+                    s
+                            .forEach(f -> sb.append(System.lineSeparator()).append("\tat ").append(f));
+                    return null;
+                });
+                System.out.println(sb.toString());
+            }
+            case "permit" -> {}
+            default -> throwIllegalAccessError(foreignAccess);
+        }
+    }
+
+    private static void throwIllegalAccessError(String value) {
+        throw new IllegalAccessError("Can not access jdk.incubator.foreign.Foreign." +
+                " System property 'jdk.incubator.foreign.Foreign' is set to '" + value + "'");
+    }
```

Comment:
The message string here should be updated to talk about the new property/method (they still talk about `Foreign`). Might also be nice to pass a string for the source of the check to checkRestrictedAccess so that it can be used in the exception message.